File: OneSignalSDK/onesignal/src/main/java/com/onesignal/language/LanguageProviderDevice.java
package com.onesignal.language;

import android.support.annotation.NonNull;

import java.util.Locale;

public class LanguageProviderDevice implements LanguageProvider {
    private static final String HEBREW_INCORRECT = "iw";
    private static final String HEBREW_CORRECTED = "he";
    private static final String INDONESIAN_INCORRECT = "in";
    private static final String INDONESIAN_CORRECTED = "id";
    private static final String YIDDISH_INCORRECT = "ji";
    private static final String YIDDISH_CORRECTED = "yi";
    private static final String CHINESE = "zh";

    @NonNull
    public String getLanguage() {
        String language = Locale.getDefault().getLanguage();

        switch (language) {
            // https://github.com/OneSignal/OneSignal-Android-SDK/issues/64
            case HEBREW_INCORRECT:
                return HEBREW_CORRECTED;
            case INDONESIAN_INCORRECT:
                return INDONESIAN_CORRECTED;
            case YIDDISH_INCORRECT:
                return YIDDISH_CORRECTED;
            // https://github.com/OneSignal/OneSignal-Android-SDK/issues/98
            case CHINESE:
                return language + "-" + Locale.getDefault().getCountry();
            default:
                return language;
        }
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/core/internal/language/LanguageProviderDevice.kt
package com.onesignal.core.internal.language

import java.util.Locale

internal class LanguageProviderDevice : ILanguageProvider {
    override val language: String
        get() {
            return when (val language = Locale.getDefault().language) {
                HEBREW_INCORRECT -> HEBREW_CORRECTED
                INDONESIAN_INCORRECT -> INDONESIAN_CORRECTED
                YIDDISH_INCORRECT -> YIDDISH_CORRECTED
                CHINESE -> language + "-" + Locale.getDefault().country
                else -> language
            }
        }

    companion object {
        private const val HEBREW_INCORRECT = "iw"
        private const val HEBREW_CORRECTED = "he"
        private const val INDONESIAN_INCORRECT = "in"
        private const val INDONESIAN_CORRECTED = "id"
        private const val YIDDISH_INCORRECT = "ji"
        private const val YIDDISH_CORRECTED = "yi"
        private const val CHINESE = "zh"
    }
}
